google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmCloudType.java

```diff
@@ -196,6 +204,56 @@ public String getPresentableName() {
       public void setName(String name) {
         this.name = name;
       }
+
+      @NotNull
+      @Override
+      public DeploymentSourceType<?> getType() {
+        return DeploymentSourceType.EP_NAME.findExtension(UserSpecifiedPathDeploymentSourceType.class);
+      }
+    }
+  }
+
+  public static class UserSpecifiedPathDeploymentSourceType extends
+      DeploymentSourceType<ModuleDeploymentSource> {
+    private static final String NAME_ATTRIBUTE = "name";
+
+    public UserSpecifiedPathDeploymentSourceType() {
+      super("filesystem WAR/JAR module");
+    }
+
+    @NotNull
+    @Override
+    /**
+     * Restore presentable name (e.g., to be "Filesystem JAR or WAR file - <file path>") of
+     * UserSpecifiedPathDeploymentSource.
+     */
+    public ModuleDeploymentSource load(@NotNull Element tag, @NotNull Project project) {
+      UserSpecifiedPathDeploymentSource userSpecifiedSource = new UserSpecifiedPathDeploymentSource(
+          ModulePointerManager.getInstance(project).create(tag.getAttributeValue(NAME_ATTRIBUTE)));
+
+      Element settings = tag.getChild(DeployToServerRunConfiguration.SETTINGS_ELEMENT);
+      if (settings != null) {
+        String filePath = settings.getAttributeValue("userSpecifiedArtifactPath");
```

Comment:
On Tue, Mar 29, 2016 at 2:23 PM, Chanseok Oh <notifications@github.com>
wrote:

> In
> google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/ManagedVmCloudType.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/548#discussion_r57774043>
> :
>
> > +      super("filesystem WAR/JAR module");
> > +    }
> > +
> > +    @NotNull
> > +    @Override
> > +    /**
> > +     * Restore presentable name (e.g., to be "Filesystem JAR or WAR file - <file path>") of
> > +     * UserSpecifiedPathDeploymentSource.
> > +     */
> > +    public ModuleDeploymentSource load(@NotNull Element tag, @NotNull Project project) {
> > +      UserSpecifiedPathDeploymentSource userSpecifiedSource = new UserSpecifiedPathDeploymentSource(
> > +          ModulePointerManager.getInstance(project).create(tag.getAttributeValue(NAME_ATTRIBUTE)));
> > +
> > +      Element settings = tag.getChild(DeployToServerRunConfiguration.SETTINGS_ELEMENT);
> > +      if (settings != null) {
> > +        String filePath = settings.getAttributeValue("userSpecifiedArtifactPath");
>
> Seems like I can't do
>
> @Attribute <https://github.com/Attribute>
> (GctBundle.message("appengine.flex.config.user.specified.artifact.attribute"))
>
woah. Why would we be localizing a persistent property name?

> in 'ManagedVmDeploymentConfiguration.java.' The error is that the
> attribute value must be constant. I think it makes more sense to make them
> final constants here actually. PTAL.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/548/files/c2dd708bb7ff20b0b6dbf3138b0d4d7c69ef1f42#r57774043>
>